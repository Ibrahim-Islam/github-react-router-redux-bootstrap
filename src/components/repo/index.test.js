import React from 'react'
import ReactDOM from 'react-dom'
import { configure, shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'
import { BrowserRouter } from 'react-router-dom'
import {Repo} from './index'

configure({ adapter : new Adapter() })

const match = { params: { username: 'test' } }
const repos = [
    {
        id: 1,
        name: 'some-project',
        link: 'https://github.com',
        desc: 'mock desc',
        lang: 'js',
        date: 'date'
    }
]
const pageInfo = {
    endCursor: '',
    hasNextPage: true
}

describe('<Repo />', () => {
    
    it('should render without crashing', () => {
        const div = document.createElement('div')
        const mockGetRepos = jest.fn()
        ReactDOM.render(
        <BrowserRouter>
            <Repo error={''} 
                repoContainer={ {repos: []} }
                match={ match }
                getRepos={mockGetRepos} />
        </BrowserRouter>, div)
        expect(mockGetRepos).toHaveBeenCalledTimes(1)
        ReactDOM.unmountComponentAtNode(div)
    })
    
    it('should match snapshot', () => {
        const mockGetRepos = jest.fn()
        const wrapper = shallow(
        <Repo error={''} 
            repoContainer={ {repos} }
            match={ match }
            getRepos={mockGetRepos} />
        )
        expect(toJson(wrapper)).toMatchSnapshot()
        expect(mockGetRepos).toHaveBeenCalledTimes(1)
        expect(wrapper.find('Table').length).toBe(1)
    })

    it('should show alert if has error', () => {
        const mockGetRepos = jest.fn()
        const wrapper = mount(
            <BrowserRouter>
                <Repo error={ 'error' } 
                    repoContainer={ { repos } }
                    match={ match }
                    getRepos={ mockGetRepos } />            
            </BrowserRouter>
        )
        expect(wrapper.find('Alert').length).toBe(1)
        expect(wrapper.find('Table').length).toBe(0)
        expect(wrapper.find('div#preloader').length).toBe(0)
        wrapper.unmount()
    })

    it('should show loading', () => {
        const mockGetRepos = jest.fn()
        const wrapper = mount(
        <BrowserRouter>
            <Repo error={ 'error' } 
                repoContainer={ { repos } }
                match={ match }
                getRepos={ mockGetRepos }
                showLoading={ true } />
        </BrowserRouter>
        )
        expect(wrapper.find('div#preloader').length).toBe(1)
        wrapper.unmount()
    })

    it('should get more repos if scrolled to bottom', () => {
        const mockGetRepos = jest.fn()
        
        const map = {};
        window.addEventListener = jest.fn((event, cb) => {
            map[event] = cb;
        });
        
        const wrapper = mount(<BrowserRouter>
            <Repo error={ 'error' } 
                repoContainer={ { repos, pageInfo } }
                match={ match }
                getRepos={ mockGetRepos }
                showLoading={ false } />
        </BrowserRouter>)

        global.innerHeight = 0
        map.scroll()
        
        expect(mockGetRepos).toHaveBeenCalledTimes(2)
        wrapper.unmount()
    })


})