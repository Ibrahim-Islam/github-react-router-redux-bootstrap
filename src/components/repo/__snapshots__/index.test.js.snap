// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Repo /> should match snapshot 1`] = `
<div>
  <div
    className="gh-table-body"
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="col-lg-12"
        >
          <Breadcrumb
            aria-label="breadcrumb"
            listTag="ol"
            tag="nav"
          >
            <BreadcrumbItem
              tag="li"
            >
              <Link
                replace={false}
                to="/"
              >
                <strong>
                  ‚Üê
                </strong>
              </Link>
            </BreadcrumbItem>
            <BreadcrumbItem
              active={true}
              tag="li"
            >
              Repositories
            </BreadcrumbItem>
          </Breadcrumb>
          <Table
            hover={true}
            responsiveTag="div"
            tag="table"
          >
            <tbody>
              <RepoItem
                date="date"
                desc="mock desc"
                id={1}
                key="1"
                lang="js"
                link="https://github.com"
                name="some-project"
              />
            </tbody>
          </Table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
